### PR TITLE
 [feat] ShardedOptim: Distributed Grad Scaler (for torch AMP) 

### LINKED ISSUE #180: [OSS] Handle torch AMP auto skip inf/nan steps
## 🚀 Feature
Using torch amp can lead to step() being skipped in the presence of inf or nans. An issue with that with OSS is that each rank will see a different step, meaning that some ranks could skip and others try to step (typically vanishing gradient on the way up and underflow with fp16). We need to handle that properly

## Motivation
Without that using torch amp with OSS can easily deadlock, see for instance https://github.com/facebookresearch/fairscale/pull/179

## Pitch
A somewhat elegant and not too far reaching solution to that could be to add a config flag to OSS, following which each rank will check for inf/nan on its grads, and signal all ranks that they should skip step() in that case. Users willing to apply OSS and torch amp would be advised to switch this flag on to prevent deadlocks

## Alternatives
Prevent mixed precision and OSS, but that makes OSS a lot less compelling

## Additional context
#179 and https://pytorch.org/docs/stable/notes/amp_examples.html?highlight=mixed%20precision#id2

### PR DESCRIPTION
**Pull request by blefaudeux**
# Before submitting

- [x] Was this discussed/approved via a Github issue? (no need for typos, doc improvements)
- [x] Did you read the [contributor guideline](https://github.com/facebookresearch/fairscale/blob/master/CONTRIBUTING.md)?
- [x] Did you make sure to update the docs?
- [x] Did you write any new necessary tests?

## What does this PR do?
Fixes #180, same as  https://github.com/facebookresearch/fairscale/pull/181 but properly rebased.
The idea is to supersede gradscaler, so that the is_finite check to prevent underflow takes the distributed state into account 

Credit: Most of the code or code ideas are from Sean (@SeanNaren)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by blefaudeux** [comment]
cc @mannatsingh

File: benchmarks/oss.py [hunk 1]
```diff
@@ -166,16 +170,25 @@ def closure():
...
+                        if scaler is not None:
+                            final_loss = closure(grad_scaler=scaler)  # AMP scaler.step does not support closures
+                            scaler.step(optimizer)  # type: ignore
+                            scaler.update()  # type: ignore
+                        else:
+                            final_loss = optimizer.step(closure)
 
-                if rank == 0:
-                    prof.export_chrome_trace(f"{optim_type}_trace.json")
+                    if rank == 0:
```
Comment (msbaines) [inline benchmarks/oss.py:180]: Why move this in?

File: benchmarks/oss.py [hunk 2]
```diff
@@ -166,16 +170,25 @@ def closure():
                 logging.info("Profiling the run")
                 with profiler.profile(use_cuda=True, record_shapes=True, profile_memory=True) as prof:  # type: ignore
                     with profiler.record_function("batch"):
-                        final_loss = optimizer.step(closure)
-                        logging.info("profiling done")
+                        if scaler is not None:
+                            final_loss = closure(grad_scaler=scaler)  # AMP scaler.step does not support closures
+                            scaler.step(optimizer)  # type: ignore
```
Comment (msbaines) [inline benchmarks/oss.py:175]: Add to stubs.

**Review by msbaines — COMMENTED** [review]
LGTM. Some minor nits.

File: fairscale/optim/grad_scaler.py [hunk 3]
```diff
@@ -3,15 +3,32 @@
...
+
 
 class GradScaler(TorchGradScaler):
     def _unscale_grads_(
         self, optimizer: Optimizer, inv_scale: Tensor, found_inf: Tensor, allow_fp16: bool
     ) -> Dict[device, Tensor]:
         return super()._unscale_grads_(optimizer, inv_scale, found_inf, True)
+
+
+class ShardedGradScaler(TorchGradScaler):
```
Comment (msbaines) [inline fairscale/optim/grad_scaler.py:23]: doc string

File: benchmarks/oss.py (hunk 1 above)
Comment (blefaudeux) [reply benchmarks/oss.py:180]: mistake indeed, my bad

**Review by msbaines — COMMENTED** [review]
Please update docs and tutorials also.

**Comment by blefaudeux** [comment]
> Please update docs and tutorials also.

good point, I completely forgot about that. Updating

File: docs/source/tutorials/oss.rst [hunk 4]
```diff
@@ -104,5 +104,45 @@ The above `train` function will then need to be run via a `multiprocessing.spawn
...
 
+Using PyTorch Automatic Mixed Precision is possible, but it requires a shard-aware GradScaler, which is available in
+`fairscale.optim.grad_scaler`. Autocast can be used as is, and the loss will be scaled and handled in the same way.
+See [the original documentation] (https://pytorch.org/docs/stable/notes/amp_examples.html?highlight=automatic%20mixed%20precision)
+for more information.
+
+.. code-block:: python
+
+
+
```
Comment (min-xu-ai) [inline docs/source/tutorials/oss.rst:117]: why 3 blank lines?

File: fairscale/optim/grad_scaler.py [hunk 5]
```diff
@@ -3,15 +3,42 @@
...
-        self, optimizer: Optimizer, inv_scale: Tensor, found_inf: Tensor, allow_fp16: bool
-    ) -> Dict[device, Tensor]:
+        self, optimizer: Optimizer, inv_scale: torch.Tensor, found_inf: torch.Tensor, allow_fp16: bool
+    ) -> Dict[torch.device, torch.Tensor]:
         return super()._unscale_grads_(optimizer, inv_scale, found_inf, True)
+
+
+class ShardedGradScaler(TorchGradScaler):
+    """
+    A shard-aware :class:`GradScaler<torch.cuda.amp.GradScaler>`, to be used in conjunction with
```
Comment (min-xu-ai) [inline fairscale/optim/grad_scaler.py:25]: add this to the API doc as well?

File: fairscale/optim/grad_scaler.py [hunk 6]
```diff
@@ -3,15 +3,42 @@
...
+        super().__init__()
+
+    def step(self, optimizer: Optimizer, *args: Any, **kwargs: Any) -> Optional[float]:
+        assert isinstance(optimizer, OSS), "ShardedGradScaler is to be used in combination with a sharded optimizer"
+
+        # Re-use the GradSCaler machinery, but make sure that the status is sync'ed in between the ranks
+        optimizer_state = self._per_optimizer_states[id(optimizer)]
+
+        for v in optimizer_state["found_inf_per_device"].values():
+            dist.all_reduce(v)
```
Comment (min-xu-ai) [inline fairscale/optim/grad_scaler.py:42]: should we do this async if it is not already?

**Review by min-xu-ai — APPROVED** [review]
very cool. some minor questions.

File: docs/source/tutorials/oss.rst (hunk 4 above)
Comment (blefaudeux) [reply docs/source/tutorials/oss.rst:117]: it popped up when `make`-ing the doc if I remember correctly, I can check that again

File: fairscale/optim/grad_scaler.py (hunk 5 above)
Comment (blefaudeux) [reply fairscale/optim/grad_scaler.py:25]: done, good point

File: fairscale/optim/grad_scaler.py (hunk 6 above)
Comment (blefaudeux) [reply fairscale/optim/grad_scaler.py]: done
